npe2/_plugin_manager.py

```diff
@@ -67,6 +75,18 @@ def discover(self):
                         self._readers[pattern].append(reader)
                     if reader.accepts_directories:
                         self._readers[""].append(reader)
+                for writer in mf.contributes.writers or []:
```

Comment:
this would be  an annoying writer... but what about a writer that supports '.tif', but only for images, and '.csv', but only for points?  they might have filename_extensions = ['.csv', ',tif'] and layer_types = ['image', 'points'] right?  I know that ideally, that would be two separate writers, but i guess this is part of the complication with writers ... we need to have a very clear idea about what a "unit" writer plugin is?  Either we need to somehow explicitly prohibit that sort of thing, or we need to restrict the spec somehow to a single extension per group of layer types?

Reply:
kinda thinking out loud here... but I wonder whether this couldn't be a single dict:

```py
self._writers: DefaultDict[Tuple[FrozenSet[str], str], List] = DefaultDict(list)
for writer in mf.contributes.writers:
    layer_types = frozenset(writer.layer_types)
    for ext in writer.file_name_extensions:
        self._writers[(layer_types, ext)].append(writer)

# example
# self._writers = {
#    (frozenset({'points', 'image'}), '.xyz'): [..., ]
# }

# then later

for (types, ext), writers in self._writers.items():
    # maybe issubset... maybe == ... that's another discussion
    # and relates to my "STORM.xyz" comment below
    if set(layer_types).issubset(types) and extension == ext:
        yield from writers

```

Reply:
I believe this also gets solved by what I just pushed.  You setup two writers that refer to the same command:

```yaml
writers:
  - command: my_plugin.writer
    filename_extensions: ['.tif']
    layer_types: ['image+']
  - command: my_plugin.writer
    filename_extensions: ['.csv']
    layer_types: ['points+']
```

In the file dialog, I believe this would end up being two separate rows in the format selection combo-box.